modules/ipp/ipp.go

```diff
@@ -3,74 +3,119 @@ package ipp
 import (
 	"bytes"
 	"encoding/binary"
-	//"io"
-	"net"
+	"errors"
+	"strings"
 )
 
-type Connection struct {
-	Conn net.Conn
-}
-
-//func ReadResponse(body *io.ReadCloser) *ScanResults {
-//	result := &ScanResults{}
-//
-//}
-
-// Returns a byte-encoded "attribute-with-one-value" with the provided "value-tag", "name", and "value"
+// Writes an "attribute-with-one-value" with the provided "value-tag", "name", and "value" to provided buffer
 // attribute-with-one-value encoding described at https://tools.ietf.org/html/rfc8010#section-3.1.4
 // Example (runnable from ipp_test.go):
 //   Input: 0x47, "attributes-charset", "us-ascii"
 //   Output: [71 0 18 97 116 116 114 105 98 117 116 101 115 45 99 104 97 114 115 101 116 0 8 117 115 45 97 115 99 105 105]
-// TODO: Should return an error when fed an invalid valueTag?
-// TODO: Determine whether this should remain public. Currently is for Testable Example
-func AttributeByteString(valueTag byte, name string, value string) []byte {
+// TODO: Switch output and Example function to use hex.Dump()
+// TODO: Should return an error when fed an invalid valueTag
+func AttributeByteString(valueTag byte, name string, value string, target *bytes.Buffer) error {
 	//special byte denoting value syntax
-	b := []byte{valueTag}
+	binary.Write(target, binary.BigEndian, valueTag)
 
-	//append 16-bit signed int denoting name length
-	l := new(bytes.Buffer)
-	binary.Write(l, binary.BigEndian, int16(len(name)))
-	b = append(b, l.Bytes()...)
+	if len(name) < (1 << 16) {
+		//append 16-bit signed int denoting name length
+		binary.Write(target, binary.BigEndian, int16(len(name)))
 
-	//append name
-	b = append(b, []byte(name)...)
+		//append name
+		binary.Write(target, binary.BigEndian, []byte(name))
+	} else {
+		return errors.New("Name too long to encode.")
+	}
 
-	//append 16-bit signed int denoting value length
-	l = new(bytes.Buffer)
-	binary.Write(l, binary.BigEndian, int16(len(value)))
-	b = append(b, l.Bytes()...)
+	if len(value) < (1 << 16) {
+		//append 16-bit signed int denoting value length
+		binary.Write(target, binary.BigEndian, int16(len(value)))
 
-	//append value
-	b = append(b, []byte(value)...)
-	return b
+		//append value
+		binary.Write(target, binary.BigEndian, []byte(value))
+	} else {
+		return errors.New("Value too long to encode.")
+	}
+
+	return nil
 }
 
+func convertURIToIPP(uri string) string {
+	if strings.HasPrefix(uri, "http://") || strings.HasPrefix(uri, "https://") {
+		uri = strings.Replace(uri, "http", "ipp", 1)
+	}
+	// TODO: RFC claims that literal IP addresses are not valid uri's, but Wireshark IPP Capture example uses them
+	// (Source: https://wiki.wireshark.org/SampleCaptures?action=AttachFile&do=view&target=ipp.pcap)
+	sections := strings.Split(uri, "/")
```

Comment:
Future -- look into `net.url.Parse()` / `URL.String()` for parsing / replacing / re-encoding URL fields?